drivers: spi: Add basic Gecko SPI driver

chrta opened the pull request:
This pull request adds SPI driver and its bindings using the USART peripheral for Silicon Labs EFM32 and EFR32 MCUs. It is a very basic driver which uses polling.

To actually use this driver, the SPI interface to the on-board flash MX25R8035F on the board efr32mg_sltb004a is also added.

The driver was tested by running the sample `samples/drivers/spi_flash`.

chrta committed:
drivers: spi: Add Gecko SPI driver

This commit adds SPI driver and its bindings using the USART peripheral
for Silicon Labs EFM32 and EFR32 MCUs.

Signed-off-by: Christian Taedcke <hacking@taedcke.com>

chrta committed:
boards: efr32mg_sltb004a: Add support for spi nor flash

This commit adds support for the on-board flash MX25R8035F that is
directly connected to the efr32mg soc.

Signed-off-by: Christian Taedcke <hacking@taedcke.com>

tbursztyka commented on drivers/spi/spi_gecko.c:
```diff
@@ -0,0 +1,329 @@
+	spi_context_update_tx(&data->ctx, 1, 1);
+
+	rx_frame = spi_gecko_recv(usart);
+	if (spi_context_rx_buf_on(&data->ctx)) {
```
tiny style issue, get an empty line before the if (it's valid if only the previously assigned variable is tested in the if)

chrta replied:
done

dbkinder commented on boards/arm/efr32mg_sltb004a/doc/index.rst:
```diff
@@ -74,6 +74,8 @@ The efr32mg_sltb004a board configuration supports the following hardware feature
 +-----------+------------+-------------------------------------+
 | I2C       | on-chip    | i2c port-polling                    |
 +-----------+------------+-------------------------------------+
+| SPI(M)    | on-chip    | spi port-polling                    |
```
What's the significance of the ``(M)`` notation?  Master? Can this just be ``SPI`` or if the ``M`` is important is should be explained?

chrta replied:
Yes, it means master. Slave-mode is not implemented by this driver. I copied the notation from https://docs.zephyrproject.org/latest/boards/arm/nrf52840_pca10059/doc/index.html#supported-features which contains `SPI (M/S)`.
I do not know what the preferred way for this is. I can also remove the `(M)`.

dbkinder replied:
If you think technical folks will know what this means, then it's OK by me.  Otherwise, maybe say ```SPI Master`` to be clear.

pabigot reviewed (commented):
I get the following with builds.  Seems like some extra checking that isn't quite right in this case (@ulfalizer?).

```
-- Loading /mnt/devel/external/zp/zephyr/boards/arm/efr32mg_sltb004a/efr32mg_sltb004a.dts as base
efr32mg_sltb004a.dts.pre.tmp:73.26-81.5: Warning (spi_bus_bridge): /soc/usart@40010800: node name for SPI buses should be 'spi'
  also defined at efr32mg_sltb004a.dts.pre.tmp:280.9-298.3
efr32mg_sltb004a.dts_compiled: Warning (spi_bus_reg): Failed prerequisite 'spi_bus_bridge'
```

If you go into `samples/drivers/spi_flash/boards` and copy one of the existing conf files to `efr32mg_sltb004a.conf` this will verify the flash.  It works for me.

I also added configuration to `tests/subsys/fs/littlefs` and confirmed to my satisfaction that it would work, except that this flash is too small to support the three test partitions the test assumes.  Once this goes in I can update it to skip the large partition.

mnkp commented on drivers/spi/spi_gecko.c:
```diff
@@ -0,0 +1,330 @@
+#define USART_PREFIX cmuClock_USART
+#define CLOCK_ID_PRFX2(prefix, suffix) prefix##suffix
+#define CLOCK_ID_PRFX(prefix, suffix) CLOCK_ID_PRFX2(prefix, suffix)
+#define CLOCK_USART(id) CLOCK_ID_PRFX(USART_PREFIX, id)
```
```
#define CLOCK_USART(id) _CONCAT(USART_PREFIX, id)
```

mnkp commented on drivers/spi/spi_gecko.c:
```diff
@@ -0,0 +1,330 @@
+	config->base->ROUTELOC1 = _USART_ROUTELOC1_RESETVALUE;
+
+	config->base->ROUTEPEN = USART_ROUTEPEN_RXPEN | USART_ROUTEPEN_TXPEN |
+	    USART_ROUTEPEN_CLKPEN;
```
For `ROUTELOC0` we use two tabs for the alignment, here one tab and spaces. Should be the same.

mnkp commented on drivers/spi/spi_gecko.c:
```diff
@@ -0,0 +1,330 @@
+
+#ifdef DT_INST_0_SILABS_GECKO_USART_SPI_LABEL
+
+SPI_INIT(0);
```
Since we have a macro already we could add a few more instances. efr32mg12 has 4 USART modules.

mnkp reviewed (commented):
@chrta Thanks, looks good with some minor comments. If you are willing to add support also for efr32_slwstk6061a board I will be happy to test it.

chrta committed:
boards: efr32_slwstk6061a: Add support for spi nor flash

This commit adds support for the on-board flash MX25R8035F that is
directly connected to the efr32fg soc.

Signed-off-by: Christian Taedcke <hacking@taedcke.com>

chrta commented:
@mnkp Thanks for your review. I addressed all your comments and also added (untested) support for the efr32_slwstk6061a board.

tbursztyka reviewed (approved):
Looks ok to me on driver side

mnkp reviewed (approved):
@chrta Thanks, works very well! Verified on efr32_slwstk6061a board.

mnkp commented:
Actually, I have one more comment. Since [Default board configuration](https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#default-board-configuration) guidelines generally require to configure all major components available on the evaluation board maybe we should ensure that spi_nor driver is enabled if flash subsystem is enabled. We could add respective section to the board configuration file, as in boards/arm/efr32_slwstk6061a/Kconfig.defconfig. Doing it manually is a bit tedious and may not be obvious to the new users.

chrta commented:
> Doing it manually is a bit tedious and may not be obvious to the new users.

I agree. Since no other board is doing it, i did not do it either. I would now add this to the Kconfig.defconfig of the two efr boards, ok @mnkp ?

```
if SPI_GECKO

config SPI_NOR
       default y

endif # SPI_GECKO

```

mnkp commented:
Hmm, I was thinking about making it dependent on `FLASH` Kconfig symbol, not `SPI_GECKO` since we want to have it available for any application that requires flash. However, after playing a bit with Kconfig I was reminded that this is not going to work. Nor reliably anyway. The problem with Kconfig.defconfig is that the default value will be taken from the first option that gets processed. From the Kconfig manual:

"A config option can have any number of default values. If multiple
  default values are visible, only the first defined one is active."

The above works at the moment but is very likely to break if someone assigns default n to `SPI_NOR` in the future in another file. The Kconfig.defconfig works well if we need to enable submodule based on the module but not the other way round.

Maybe we could use `boards/arm/efr32_slwstk6061a/efr32_slwstk6061a_defconfig` and have it always enabled? We already configure SPI there, we would also need to enable FLASH.

Unless someone has a better idea?

chrta commented:
@mnkp I would prefer to leave it as it is now and create an new github issue for the discussion. Since this should be done not only for the Silabs boards, but also for nordic boards etc.

mnkp commented:
> I would prefer to leave it as it is now and create an new github issue for the discussion.

Sounds good.

One more comment. The commits in #20181 got merged into master so I think it would make sense if this PR was reverted to target master as well. There are no dependencies on `topic-gpio` branch anymore.

chrta commented:
This pull request now targets master branch. I also cleaned up the reviewers that were added automatically by github .

mnkp commented on boards/arm/efr32_slwstk6061a/efr32_slwstk6061a_defconfig:
```diff
@@ -7,6 +7,7 @@ CONFIG_BOARD_EFR32_SLWSTK6061A=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
 CONFIG_SERIAL=y
+CONFIG_SPI=y
```
Despite "Default Board Configuration" guidelines requirement to enable all major components on the evaluation board there are sample applications, like `samples/basic/minimal/` which expect default configuration not to enable drivers which are not required. Maybe we should remove this option from `_defconfig` files. Sample application that require SPI enable `CONFIG_SPI` in their respective `prj.conf` file.

chrta replied:
done, but i noted that a few other boards (about 10) also enable spi in their board config.

mnkp replied:
Indeed, the approach to configuring peripherals is still a bit inconsistent.

chrta commented:
@galak Is this now ready to merge?

MaureenHelm reviewed (changes requested):
please fix the checkpatch warnings

chrta commented:
> please fix the checkpatch warnings

This is the code that produces the warning:
```c
#ifndef CONFIG_SOC_GECKO_HAS_INDIVIDUAL_PIN_LOCATION
#error "This EFM32 USART SPI driver is only implemented for devices that \
 support individual pin locations"
#endif
```

Could you suggest a fix. I could also write "Not implemented for this SoC", but it is less detailed.

MaureenHelm commented:
> > please fix the checkpatch warnings
> 
> This is the code that produces the warning:
> 
> ```c
> #ifndef CONFIG_SOC_GECKO_HAS_INDIVIDUAL_PIN_LOCATION
> #error "This EFM32 USART SPI driver is only implemented for devices that \
>  support individual pin locations"
> #endif
> ```
> 
> Could you suggest a fix. I could also write "Not implemented for this SoC", but it is less detailed.

How about: "Individual pin location support is required"

chrta commented:
@MaureenHelm Changed error message to your suggestion.

chrta commented:
Shippable failed because of some unrelated failed test. I also rebased this pr onto the current master. The failed test:
```
2020-01-10 09:36:23,730 - INFO - 814/814 qemu_x86_64               tests/lib/mem_alloc/libraries.libc.minimal.mem_alloc FAILED Exited with 1 (qemu)
```
